### PR TITLE
Playwright Utils: Fix the method of getting post ID in 'publishPost'

### PR DESCRIPTION
**Pull request by Mamaduka**
## What?
This is a follow-up to https://github.com/WordPress/gutenberg/pull/56253#discussion_r1397306664.

PR fixes the method of getting post ID in `publishPost` to avoid failures when the post publish panel isn't displayed.

## Why?
Displaying the post-publish panel isn't guaranteed. It can be hidden when a post-type isn't viewable or disabled using the "Include pre-publish checklist" preference.

## How?
* Wait for the "Post Published" notice instead of the panel.
* Get the post ID from the editor's page URL.

## Testing Instructions
Playwright E2E tests are passing on CI.

### CODE REVIEW

File: test/e2e/specs/editor/various/sidebar-permalink.spec.js [hunk 1]
```diff
@@ -48,7 +48,7 @@ test.describe( 'Sidebar Permalink', () => {
 		await editor.canvas
 			.getByRole( 'textbox', { name: 'Add title' } )
 			.fill( 'aaaaa' );
-		await editor.saveDraft();
+		await editor.publishPost();
```
Comment (WunderBart) [inline test/e2e/specs/editor/various/sidebar-permalink.spec.js:51]: Why this change here?

File: packages/e2e-test-utils-playwright/src/editor/publish-post.ts [hunk 2]
```diff
@@ -28,12 +28,11 @@ export async function publishPost( this: Editor ) {
...
-		.inputValue();
-	const url = new URL( urlString );
-	const postId = url.searchParams.get( 'p' );
+	await this.page
+		.getByRole( 'button', { name: 'Dismiss this notice' } )
+		.filter( { hasText: 'published' } )
+		.waitFor();
+	const postId = new URL( this.page.url() ).searchParams.get( 'post' );
 
 	return typeof postId === 'string' ? parseInt( postId, 10 ) : null;
```
Comment (WunderBart) [inline packages/e2e-test-utils-playwright/src/editor/publish-post.ts:37]: In which situations would the `postId` be unavailable?

File: packages/e2e-test-utils-playwright/src/editor/publish-post.ts [hunk 3]
```diff
@@ -28,12 +28,11 @@ export async function publishPost( this: Editor ) {
...
-		.getByRole( 'region', { name: 'Editor publish' } )
-		.getByRole( 'textbox', { name: 'address' } )
-		.inputValue();
-	const url = new URL( urlString );
-	const postId = url.searchParams.get( 'p' );
+	await this.page
+		.getByRole( 'button', { name: 'Dismiss this notice' } )
+		.filter( { hasText: 'published' } )
+		.waitFor();
+	const postId = new URL( this.page.url() ).searchParams.get( 'post' );
```
Comment (WunderBart) [inline packages/e2e-test-utils-playwright/src/editor/publish-post.ts:35]: This looks OK, but if we expect the `post` param to be there, we might want to wrap it with `waitForFunction` or an `expect.poll` so we're not fetching it too early.

File: test/e2e/specs/editor/various/sidebar-permalink.spec.js (hunk 1 above)
Comment (Mamaduka) [reply test/e2e/specs/editor/various/sidebar-permalink.spec.js:51]: Mostly to showcase that it works. See https://github.com/WordPress/gutenberg/pull/56253#discussion_r1397306664.

File: packages/e2e-test-utils-playwright/src/editor/publish-post.ts (hunk 3 above)
Comment (Mamaduka) [reply packages/e2e-test-utils-playwright/src/editor/publish-post.ts:35]: I don't have strong opinions here, but if the query param isn't available, there's something wrong with the test, and it's better to fail.

We're already using a similar pattern without any additional checks. Here's an example from perf utils:

https://github.com/WordPress/gutenberg/blob/a7911582cda6f11d1ebddcd27ccbb82fb86022c5/test/performance/fixtures/perf-utils.ts#L62

P.S. The `post` query arg doesn't change after a post is published or saved as a draft.

File: packages/e2e-test-utils-playwright/src/editor/publish-post.ts (hunk 2 above)
Comment (Mamaduka) [reply packages/e2e-test-utils-playwright/src/editor/publish-post.ts:37]: Not sure. The safeguard was here, so I just left it.

File: packages/e2e-test-utils-playwright/src/editor/publish-post.ts (hunk 3 above)
Comment (WunderBart) [reply packages/e2e-test-utils-playwright/src/editor/publish-post.ts:35]: > (…) if the query param isn't available, there's something wrong with the test, and it's better to fail.

That's why [I asked](https://github.com/WordPress/gutenberg/pull/56421#discussion_r1401871321) if there's a situation where the post id would not be available after saving/publishing. If not, we should probably throw from this util instead of returning `null`, IIUC.


> The post query arg doesn't change after a post is published or saved as a draft.

Right, the subsequent saves would already have the `post` param. The only potential flakey would come from the first save, then. I guess this is fine, though it would not be the first time I see racing issues caused by assuming one element's state off of another's – in this case, URL off of the post-publish notice.

Comment (Mamaduka) [reply packages/e2e-test-utils-playwright/src/editor/publish-post.ts:35]: > I guess this is fine, though it would not be the first time I see racing issues caused by assuming one element's state off of another element's state

E2E tests have to make those assumptions, IMO.

> If not, we should probably throw from this util instead of returning null, IIUC.

Somethings like this?

```js
if ( postId !== 'string' ) {
  throw new Error( 'Unable to retrieve post ID from URL.' );
}

return parseInt( postId, 10 );
```

**Review by kevin940726 — APPROVED** [review]
I agree that failing fast is usually better. But given that the code was already there, I don't think it blocks this PR 👍

Nice work! Thank you!

**Comment by Mamaduka** [comment]
Thanks for the feedback! I will merge this and be happy to do a follow-up if/when there's a need for it.

File: packages/e2e-test-utils-playwright/src/editor/publish-post.ts (hunk 3 above)
Comment (WunderBart) [reply packages/e2e-test-utils-playwright/src/editor/publish-post.ts:35]: > E2E tests have to make those assumptions, IMO.

Not if we can wait for stuff! 😄 (post ID in this case)

> Somethings like this?

I realized now that we could get the ID via the `getCurrentPostId()` which we don't need to wait for (`id` is set when the `new-post.php` is open)

```js
const postId = await this.page.evaluate( () => {
	return window.wp.data.select( 'core/editor' ).getCurrentPostId();
} );

return postId;
```

Comment (WunderBart) [reply packages/e2e-test-utils-playwright/src/editor/publish-post.ts:35]: ...but we can also wait for the `post` param, like this:

```js
const postId = await this.page.waitForFunction( () => {
	const url = new URL( window.location.href );
	if ( url.searchParams.has( 'post' ) ) {
		return url.searchParams.get( 'post' );
	}

	return false;
} );

return parseInt( postId, 10 );
```

Comment (Mamaduka) [reply packages/e2e-test-utils-playwright/src/editor/publish-post.ts:35]: Thank you, @WunderBart 🙇 Do you think we should update the code or keep it as it is for now?

Comment (WunderBart) [reply packages/e2e-test-utils-playwright/src/editor/publish-post.ts:35]: If it ain't broke, don't fix it, right? 😄 No pressure here - let's just keep that in mind in case it starts making trouble.

Comment (Mamaduka) [reply packages/e2e-test-utils-playwright/src/editor/publish-post.ts:35]: Sounds good!
